### PR TITLE
Clarify documentation regarding "cdef public"

### PR DESCRIPTION
**Pull request by jdemeyer**
I think the documentation about `cdef public` is unclear, especially about the "linking" part. In particular, it seems to imply that you can link code against the extension module (i.e. using the `.so` file as a dynamic library). That works on Linux, but isn't portable (it does not work on OS X for example).

### CODE REVIEW

File: docs/src/userguide/external_C_code.rst [hunk 1]
```diff
@@ -339,42 +339,57 @@ Public Declarations
 ---------------------
 
 You can make C types, variables and functions defined in a Cython module
-accessible to C code that is linked with the module, by declaring them with
-the public keyword::
+accessible to C code that is linked together with the Cython-generated C file,
```
Comment (embray) [inline docs/src/userguide/external_C_code.rst]: To be clear, "the Cython-generated C file" is not what's linked of course.  Though I'm struggling to come up with more concise wording than "the object file compiled from the Cython-generated C file"

**Comment by jdemeyer** [comment]
> To be clear, "the Cython-generated C file" is not what's linked of course

Technically, you are right. But I think this is a case where it is clear what is meant.

**Comment by embray** [comment]
I'm not sure--I think it's clear enough to us because we know what the use case is.  But I could see someone coming to this for the first time being confused.  That said, a slight change of wording wouldn't fix that either.  I think maybe the specific example you give later is good enough to clarify.

**Comment by scoder** [comment]
In any case, it's better than before. Thanks.
